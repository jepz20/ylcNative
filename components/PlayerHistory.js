// @flow

import * as React from 'react'
import {
  View,
  Text,
  StyleSheet,
  ScrollView,
  TouchableWithoutFeedback
} from 'react-native'
import type { LogMap, LogList } from '../types/match'
import { mapDueslPerMatch } from '../core'
import { NO_LOG_CHARACTER } from '../constants'
import { Map } from 'immutable'
import type { ToggleCalculatorVisibility } from '../actions/calculator'

const POSITIVE = 'green'
const NEGATIVE = 'red'
type Props = {
  logs: LogMap,
  playerId: string,
  currentDuel: string,
  toggleCalculatorVisibility: ToggleCalculatorVisibility
}

const PlayerHistory = ({
  logs,
  playerId,
  currentDuel = '1',
  toggleCalculatorVisibility
}: Props) => {
  const currentLog: ?LogList = logs.get(String(currentDuel))
  const renderLog = (currentLog: LogList) =>
    currentLog
      .reverse()
      .map(({ operationValue, playerId: logPlayerId, previousPoints, id }) => {
        if (logPlayerId !== playerId) return null
        return (
          <View style={[styles.logContainer]} key={id}>
            <Text
              style={[
                styles.operationValue,
                { color: operationValue > 0 ? POSITIVE : NEGATIVE }
              ]}
            >
              {operationValue > 0 && '+'}
              {operationValue}
            </Text>
            <Text style={[styles.currentPoints]}>{previousPoints}</Text>
          </View>
        )
      })

  const renderNoLogs = (noLogChar: string) =>
    mapDueslPerMatch(index => (
      <View style={styles.logContainer} key={index}>
        <Text style={styles.noLog}>{noLogChar}</Text>
      </View>
    ))

  return (
    <ScrollView contentContainerStyle={styles.container}>
      <TouchableWithoutFeedback
        onPress={() => toggleCalculatorVisibility(playerId)}
      >
        <View>
          {currentLog ? renderLog(currentLog) : renderNoLogs(NO_LOG_CHARACTER)}
        </View>
      </TouchableWithoutFeedback>
    </ScrollView>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'flex-start'
  },
  logContainer: {
    marginTop: 10,
    marginBottom: 10,
    flexDirection: 'row',
    justifyContent: 'center'
  },
  operationValue: {
    alignSelf: 'flex-end',
    fontSize: 15,
    marginRight: 10,
    marginBottom: 5,
    flex: 1,
    textAlign: 'right'
  },
  currentPoints: {
    fontSize: 30,
    flex: 2,
    textDecorationLine: 'line-through',
    textDecorationStyle: 'solid',
    color: 'grey'
  },
  noLog: {
    fontSize: 30,
    color: 'grey',
    alignSelf: 'flex-end'
  }
})

PlayerHistory.defaultProps = {
  logs: Map()
}

export { PlayerHistory }
