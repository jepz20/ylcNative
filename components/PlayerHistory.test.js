import React from 'react'
import { TouchableWithoutFeedback } from 'react-native'
import { Map, Record, List } from 'immutable'
import renderer from 'react-test-renderer'

import { PlayerHistory } from './PlayerHistory'

describe('PlayerHistory', () => {
  test('renders with no props ', () => {
    const playerDetail = renderer.create(<PlayerHistory />).toJSON()
    expect(playerDetail).toMatchSnapshot()
  })

  test('renders with empty logs ', () => {
    const log = Map()
    const playerDetail = renderer
      .create(
        <PlayerHistory
          playerId='1'
          currentDuel='2'
          toggleCalculatorVisibility={() => {}}
          logs={log}
        />
      )
      .toJSON()
    expect(playerDetail).toMatchSnapshot()
  })

  test('renders with first duel with no movements ', () => {
    const log = Map({
      '1': List()
    })
    const playerDetail = renderer
      .create(
        <PlayerHistory
          playerId='1'
          currentDuel='2'
          toggleCalculatorVisibility={() => {}}
          logs={log}
        />
      )
      .toJSON()
    expect(playerDetail).toMatchSnapshot()
  })

  test('renders with first duel with 2 logs ', () => {
    const logRecord = Record({
      id: '1',
      playerId: '1',
      operationValue: 10,
      currentPoints: 10,
      previousPoints: 20,
      type: 'operation'
    })

    const log = Map({
      '1': List([
        logRecord(),
        logRecord({ id: '2', playerId: '2' }),
        logRecord({ id: '3', playerId: '1', operationValue: -22 })
      ])
    })
    const playerDetail = renderer
      .create(
        <PlayerHistory
          playerId='1'
          currentDuel='1'
          toggleCalculatorVisibility={jest.fn()}
          logs={log}
        />
      )
      .toJSON()
    expect(playerDetail).toMatchSnapshot()
  })

  test('renders with second duel when no logs, when first duel has 2 logs ', () => {
    const logRecord = Record({
      id: '1',
      playerId: '1',
      operationValue: 10,
      currentPoints: 10,
      previousPoints: 20,
      type: 'operation'
    })

    const log = Map({
      '1': List([logRecord(), logRecord({ id: '2', playerId: '2' })])
    })
    const playerDetail = renderer
      .create(
        <PlayerHistory
          playerId='1'
          currentDuel='2'
          toggleCalculatorVisibility={jest.fn()}
          logs={log}
        />
      )
      .toJSON()
    expect(playerDetail).toMatchSnapshot()
  })

  test('should toggle visibility', () => {
    const logRecord = Record({
      id: '1',
      playerId: '1',
      operationValue: 10,
      currentPoints: 10,
      previousPoints: 20,
      type: 'operation'
    })

    const log = Map({
      '1': List([logRecord(), logRecord({ id: '2', playerId: '2' })])
    })

    const mockToggle = jest.fn()
    const playerDetail = renderer.create(
      <PlayerHistory
        playerId='1'
        currentDuel='2'
        toggleCalculatorVisibility={mockToggle}
        logs={log}
      />
    )
    const playerDetailInstance = playerDetail.root
    const touchable = playerDetailInstance.findByType(TouchableWithoutFeedback)
    touchable.props.onPress()
    playerDetail.root.props.toggleCalculatorVisibility()
    expect(mockToggle).toBeCalled()
  })
})
