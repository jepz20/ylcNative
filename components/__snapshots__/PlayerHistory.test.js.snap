// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerHistory renders with empty logs  1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flex": 1,
      "justifyContent": "flex-start",
    }
  }
>
  <View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={undefined}
      testID={undefined}
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "flex-end",
              "color": "grey",
              "fontSize": 30,
            }
          }
        >
          .
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "flex-end",
              "color": "grey",
              "fontSize": 30,
            }
          }
        >
          .
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "flex-end",
              "color": "grey",
              "fontSize": 30,
            }
          }
        >
          .
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`PlayerHistory renders with first duel with 2 logs  1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flex": 1,
      "justifyContent": "flex-start",
    }
  }
>
  <View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={undefined}
      testID={undefined}
    >
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "justifyContent": "center",
              "marginBottom": 10,
              "marginTop": 10,
            },
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "alignSelf": "flex-end",
                "flex": 1,
                "fontSize": 15,
                "marginBottom": 5,
                "marginRight": 10,
                "textAlign": "right",
              },
              Object {
                "color": "red",
              },
            ]
          }
        >
          -22
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "grey",
                "flex": 2,
                "fontSize": 30,
                "textDecorationLine": "line-through",
                "textDecorationStyle": "solid",
              },
            ]
          }
        >
          20
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "justifyContent": "center",
              "marginBottom": 10,
              "marginTop": 10,
            },
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "alignSelf": "flex-end",
                "flex": 1,
                "fontSize": 15,
                "marginBottom": 5,
                "marginRight": 10,
                "textAlign": "right",
              },
              Object {
                "color": "green",
              },
            ]
          }
        >
          +
          10
        </Text>
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "grey",
                "flex": 2,
                "fontSize": 30,
                "textDecorationLine": "line-through",
                "textDecorationStyle": "solid",
              },
            ]
          }
        >
          20
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`PlayerHistory renders with first duel with no movements  1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flex": 1,
      "justifyContent": "flex-start",
    }
  }
>
  <View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={undefined}
      testID={undefined}
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "flex-end",
              "color": "grey",
              "fontSize": 30,
            }
          }
        >
          .
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "flex-end",
              "color": "grey",
              "fontSize": 30,
            }
          }
        >
          .
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "flex-end",
              "color": "grey",
              "fontSize": 30,
            }
          }
        >
          .
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`PlayerHistory renders with no props  1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flex": 1,
      "justifyContent": "flex-start",
    }
  }
>
  <View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={undefined}
      testID={undefined}
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "flex-end",
              "color": "grey",
              "fontSize": 30,
            }
          }
        >
          .
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "flex-end",
              "color": "grey",
              "fontSize": 30,
            }
          }
        >
          .
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "flex-end",
              "color": "grey",
              "fontSize": 30,
            }
          }
        >
          .
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`PlayerHistory renders with second duel when no logs, when first duel has 2 logs  1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flex": 1,
      "justifyContent": "flex-start",
    }
  }
>
  <View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={undefined}
      testID={undefined}
    >
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "flex-end",
              "color": "grey",
              "fontSize": 30,
            }
          }
        >
          .
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "flex-end",
              "color": "grey",
              "fontSize": 30,
            }
          }
        >
          .
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "center",
            "marginBottom": 10,
            "marginTop": 10,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "flex-end",
              "color": "grey",
              "fontSize": 30,
            }
          }
        >
          .
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
